### PR TITLE
[NPU] Migrate async CAM to source-built routed-only operators

### PR DESCRIPTION
**Pull request by jiangkuaixue123**
## Purpose

Switch Async CAM from the legacy wheel to the four source-built `torch.ops.afd_ascend.afd_async_*` operators. Attention now owns shared experts, while FFN processes routed-only compact payloads. Preserve native DSV4 routing precision by computing gate logits with FP32 weights.

## Issue

- Builds on the source operators introduced in #361.

## Scope

- NPU eager Async CAM and existing two-stage MoE ubatching for DeepSeek-V2-Lite and DeepSeek-V4-Flash.
- Remove legacy CAM loader/vendor environment setup from runtime, E2E, launch scripts, and documentation.
- No changes to `csrc/`, upstream source checkouts, GPU execution, or new graph/DBO capabilities.

## Implementation Notes

- Require all four plugin-owned async operators at startup.
- Consume four-tensor routed-only receive results, use expert counts for valid chunk length, preserve compact metadata, and notify completion for empty ranks and multiple chunks.
- Load and compute shared experts on Attention with replicated weights and local token shards; preserve scaling, quantization, padding, stage pairing, and connector shutdown cleanup.
- Match the native DSV4 FP32 gate calculation. The checkpoint probe exposed 29 differing expert IDs with BF16 logits; FP32 makes IDs and weights match.
- Reject unsupported Async CAM `mix_placement` at startup.

## Test Plan

Run source-operator round trips, checkpoint MoE comparisons, targeted unit tests, and the existing NPU E2E entrypoints. Validate DSV2-Lite before DSV4-Flash, followed by final snapshot and resource-cleanup checks.

## Test Result

Validated on `jcz_afd1`, 16 Ascend 910C devices, CANN/HCCL 9.0.1:

- Full source build succeeded; four async operators and A2E/E2A registered. Operator sources are unchanged.
- 12/12 real-NPU round trips passed: TP1/2/4 × FP16/BF16 × dynamic quantization 0/1, including empty ranks, multiple chunks, sparse routes, decode, and reused windows.
- DSV2-Lite completion and two-stage ubatch passed with zero skips. GSM8K strict-match: 2/7 (0.2857); all seven filtered answers and prompt/target hashes match the synchronous CAMP2P reference.
- DSV2 BF16/FP16 checkpoint probes passed. DSV4 Hash layer 0 and non-Hash layer 3 probes passed with unchanged tolerances.
- DSV4 DP2×TP4 Attention + EP8 FFN, dynamicQuant=1, two stages: 10/10 concurrent responses have correct answers 19–28 and `finish_reason=stop`, zero skips.
- Targeted unit group: 152 passed, 2 skipped because the built Ascend extension is present. Final gate-selector tests: 12 passed; loader tests: 8 passed with the same 2 environment skips.
- Ruff check/format, compile checks, diff hygiene, and SPDX checks passed. Final NPU processes and test ports were released.

Review follow-up: removed the loader's informational logging, excluded the migration plan and validation report, restored the DeepSeek-V3.2 recipe README, and removed the cross-node launch script and its stale reference. The excluded documents and script are preserved with the local validation evidence.

Review validation: all 10 CPU loader tests passed. Hardware results above are from the migration validation before this review follow-up; NPU E2E was not rerun for these changes.

Evidence limits: initial build stdout was not retained; local evidence records the observed success and built-library hashes. Teardown emitted peer-close/forced-cleanup warnings after results were saved, with no remaining device processes or test listeners.

Latest review follow-up: restore `_log_cam_op_values` for the routed-only receive outputs; rename the connector capacity member to `max_num_batched_tokens` while preserving the operator's `max_seq_len` contract; restore the FFN runner's bounded loop because the outer daemon continues receiving chunks; remove redundant model-level exception cleanup and its unused work-item/helper state. Retain shared-output staging and the native FP32 DSV4 gate computation. Update the affected unit tests, including chunks crossing runner-call boundaries.

Validation of this latest follow-up is deferred at the author's request. Earlier results above do not certify the final follow-up diff; no further tests or NPU E2E are being run as part of this update.

## Docs Impact

Updated connector/FFN/model design contracts, NPU setup and troubleshooting, remaining recipes and E2E instructions. Added operator and checkpoint-reference test instructions.

---

<details>
<summary>Essential PR Checklist</summary>

- [x] Purpose is clear and linked to public context when possible.
- [x] Scope is bounded.
- [x] Compatibility with vLLM v0.26.0 is considered.
- [x] No changes are made to the vLLM source checkout.
- [x] Plugin-owned classes or explicit dotted class paths are preferred over monkey patches.
- [x] Any compat shim or monkey patch is isolated, idempotent, version-guarded, documented, and tested.
- [x] Imports remain CPU-safe; CUDA-heavy work is delayed or GPU-gated.
- [x] Validation evidence is included, including skipped GPU tests when applicable.
- [x] Documentation impact is stated.

</details>

### CODE REVIEW

File: afd_plugin/compat/npu/ops.py [hunk 1]
```diff
@@ -105,28 +108,26 @@ def has_afd_ascend_ops() -> bool:
...
     except AttributeError as exc:
         raise RuntimeError(
-            "CAMAsyncAFDConnector requires real torch.ops.umdk_cam_op_lib CAM ops "
-            "(async_dispatch_send, async_dispatch_recv, async_combine_send, "
-            "async_combine_recv). Install or preload the CAM operator binaries "
-            "before initializing the connector.",
+            "Async CAM requires all four torch.ops.afd_ascend.afd_async_* "
+            "operators. Rebuild AFD with AFD_BUILD_ASCEND_OPS=1."
         ) from exc
+    # Inherit the worker's configured vLLM handler instead of root WARNING.
```
Comment (jiangkuaixue123) [inline afd_plugin/compat/npu/ops.py]: 这个logging去掉

File: docs/npu/CAM_ASYNC_SOURCE_MIGRATION_PLAN.md [hunk 2]
```diff
@@ -0,0 +1,401 @@
+# Async CAM 源码算子适配计划
```
Comment (jiangkuaixue123) [inline docs/npu/CAM_ASYNC_SOURCE_MIGRATION_PLAN.md]: 这个文件不提交

File: docs/npu/CAM_ASYNC_SOURCE_MIGRATION_VALIDATION.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
+# CAM Async 源码算子迁移验收记录
```
Comment (jiangkuaixue123) [inline docs/npu/CAM_ASYNC_SOURCE_MIGRATION_VALIDATION.md]: 这个文件不提交

File: recipe/npu/CAMAsyncAFDConnector/deepseek_v3_2/README.md [hunk 4]
```diff
@@ -46,8 +46,7 @@ docker pull quay.io/ascend/vllm-ascend:v0.19.1rc1-a3-openeuler
 Run the following commands from the repository root inside the container:
```
Comment (jiangkuaixue123) [inline recipe/npu/CAMAsyncAFDConnector/deepseek_v3_2/README.md:46]: 去掉这个文件的所有修改

File: tools/itask/launch_dsv4_afd_cross_node.sh [hunk 5]
```diff
@@ -83,17 +83,9 @@ export PYTHONPATH="/vllm-workspace/vllm-ascend:/vllm-workspace/vllm:${PYTHONPATH
 export VLLM_PLUGINS=ascend,afd
```
Comment (jiangkuaixue123) [inline tools/itask/launch_dsv4_afd_cross_node.sh]: 移除这个文件

File: afd_plugin/connectors/npu/async_cam.py [hunk 6]
```diff
@@ -740,32 +714,10 @@ def recv_attn_output(
...
-        (
-            hidden_states,
-            expand_x_shared,
-            dynamic_scales,
-            dynamic_scales_shared,
-            token_nums_rankid_layeridx,
-            expert_token_nums,
-            expert_token_nums_shared,
-        ) = outputs
-        _log_cam_op_values(
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py:752]: 这个文件中的定位日志_log_cam_op_values不要删除

File: afd_plugin/connectors/npu/async_cam.py [hunk 7]
```diff
@@ -724,11 +698,11 @@ def recv_attn_output(
             dynamic_quant=self.dynamic_quant,
             group_name=self.group_name,
         )
-        outputs = torch.ops.umdk_cam_op_lib.async_dispatch_recv(
+        outputs = torch.ops.afd_ascend.afd_async_dispatch_recv(
             placeholder,
             self.comm_args,
             self.comm_id,
-            states.batch_size,
+            self.max_seq_len,
```
Comment (jiangkuaixue123) [inline afd_plugin/connectors/npu/async_cam.py]: 建议将 connector 内的 `self.max_seq_len` 统一重命名为 `self.max_num_batched_tokens`，并同步更新初始化和所有引用。它实际来自 `vllm_config.scheduler_config.max_num_batched_tokens`，表示调度 token 预算及 CAM 通信容量；当前名称容易与单条序列的最大上下文长度（`max_model_len`）混淆。底层算子的 `max_seq_len` 参数名可以保持不变。

File: afd_plugin/model_executor/models/npu/deepseek_v2_async_cam_forward.py [hunk 8]
```diff
@@ -81,25 +82,31 @@ def run_model_forward(
...
-        hidden_states, residual = run_async_moe_ubatch_afd_forward(
-            model,
-            hidden_states,
-            residual,
-            positions,
-            afd_metadata,
-            async_moe_ubatch_metadata,
-            llama_4_scaling,
-        )
+    try:
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/npu/deepseek_v2_async_cam_forward.py]: 为什么这里要加try

File: afd_plugin/model_executor/models/npu/deepseek_v2_async_cam_forward.py [hunk 9]
```diff
@@ -122,6 +129,7 @@ def run_attention_gate_afd_forward(
     afd_connector = afd_metadata.connector
     forward_context = get_forward_context()
     stage_idx = afd_metadata.stage_idx
+    pending_shared_output: torch.Tensor | None = None
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/npu/deepseek_v2_async_cam_forward.py:125]: 为什么要添加pending_shared_output

File: afd_plugin/v1/worker/npu/ffn_model_runner.py [hunk 10]
```diff
@@ -320,7 +320,8 @@ def _ffn_forward_connector_driven(
         rank_ffn_output = None
         connector = cast(CAMAsyncAFDConnector, self.connector)
 
-        for _ in _ffn_layer_indices(self):
+        completed_layers = 0
+        while completed_layers < len(_ffn_layer_indices(self)):
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/npu/ffn_model_runner.py]: This is no need.

File: afd_plugin/v1/worker/npu/ffn_model_runner.py [hunk 11]
```diff
@@ -360,13 +361,15 @@ def _ffn_forward_connector_driven(
...
                     group_list=states.group_list,
                     dynamic_scales=states.dynamic_scales,
-                    expand_x_shared=states.expand_x_shared,
-                    dynamic_scales_shared=states.dynamic_scales_shared,
                 )
                 rank_ffn_output = connector.send_ffn_work_item_output(
                     work_item,
                     rank_ffn_output,
                 )
+            # Receive chunks can split one layer across expert intervals.
```
Comment (jiangkuaixue123) [inline afd_plugin/v1/worker/npu/ffn_model_runner.py]: This modify is no need.

File: afd_plugin/model_executor/models/npu/deepseek_v4_attention_gate.py [hunk 12]
```diff
@@ -113,7 +113,11 @@ def compute_attention_gate_topk(
     tokens, then hand their IDs and weights to CAM dispatch.
     """
 
-    router_logits, _ = moe.gate(hidden_states)
+    # Match AscendMoERunner's internal router: BF16 GEMM can change expert
```
Comment (jiangkuaixue123) [inline afd_plugin/model_executor/models/npu/deepseek_v4_attention_gate.py:116]: 对于这个修改我有疑问
